### DISCUSSION TITLE
How can I access timestamps for when the app was built, its version, git hash, and things like that?

### DISCUSSION THREAD

**Discussion by Svish** [vercel/next.js#50181] [Help]
### Summary

Struggling to find documentation on how to access this information. I find some info, but it's either very outdated, or very hacky, or usually both.

Is there a proper, straight forward, not hacky, way to access these kinds of data in a Next app?

Some reasons I want this:

* I want to [generate a `sitemap.xml`](https://nextjs.org/docs/app/api-reference/file-conventions/metadata/sitemap#generate-a-sitemap), but need a way to generate a sensible `lastModified` date, for example the build date, or the date of the current git commit, etc.
* I want to show a link to my github repo, directly to the currently deployed version.

### Additional information

```js
Using latest NextJS and the app directory.


    Operating System:
      Platform: win32
      Arch: x64
      Version: Windows 10 Pro
    Binaries:
      Node: 20.1.0
      npm: N/A
      Yarn: N/A
      pnpm: N/A
    Relevant packages:
      next: 13.4.3
      eslint-config-next: 13.4.3
      react: 18.2.0
      react-dom: 18.2.0
      typescript: 5.0.4
```

**Comment by masterkain** [comment]
maybe overwrite a placeholder json file with the relevant data you want to load during one of your build steps and import it

**Comment by icyJoseph** [comment]
Would it make sense to use the `generateBuildId` key in the next.config.js? https://nextjs.org/docs/app/api-reference/next-config-js/generateBuildId 

There you can run a child process which would execute a command to invoke the latest git commit hash, and its date. The build id placed in `./next/BUILD_ID` - in pages directory the buildId is present in `window.__NEXT_DATA__`.

You could also have a build script, that outputs the info you need as JSON into the public directory, or as part of .env, and then it also builds the project. Again using child processes. https://nodejs.org/api/child_process.html - you have the power of Node.js at your hands, you can do as you wish.

**Reply by Svish** [reply]
I don't know how to do this, so I don't know whether this would make sense of not, that's why I'm asking🤷‍♂️

Do you know of a concrete example where something like this is done, so I can look at how the code actually looks?

**Reply by icyJoseph** [reply]
Ah, I see, sorry I made an assumption.

```js
// scripts/build.js
const { spawn, execSync } = require("child_process");

const lastCommit = "git rev-parse HEAD";

const commit = execSync(lastCommit).toString();

const child = spawn("yarn", ["build"], {
  stdio: "inherit",
  env: {
    ...process.env,
    NEXT_PUBLIC_LAST_COMMIT: commit,
  },
});

child.on("error", (error) => {
  console.log("error", error);
});

child.on("exit", () => {
  console.log({ commit });
  console.log("Done");
});
```

This one for example, makes `NEXT_PUBLIC_LAST_COMMIT` available.

So you can run `node scripts/build.js`, and for example this bit:

```jsx
export default function Home() {
  return (
    <main>
      <p>{process.env.NEXT_PUBLIC_LAST_COMMIT}</p>
    </main>
  );
}
```

Renders the SHA-1 for the latest HEAD. 

Now you can change your package.json to include a build script that runs `node scripts/build.js` and you are good to go.

![Screenshot 2023-05-23 at 13 07 13](https://github.com/vercel/next.js/assets/21013447/3948b3a6-fe3d-4aba-86b4-ec0ccb5e6406)

**Reply by kalaschnik** [reply]
I cannot get this to work. Ho do I hook this with the build script? Is there also a way to show this during development?
